File: src/java/magma/collect/stream/Stream.java
package magma.collect.stream;

import magma.option.Option;
import magma.result.Result;

import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;

public interface Stream<T> {
    <R> R foldWithInitial(R initial, BiFunction<R, T, R> folder);

    <R> Stream<R> map(Function<T, R> mapper);

    <C> C collect(Collector<T, C> collector);

    <R> Stream<R> flatMap(Function<T, Stream<R>> mapper);

    <R, X> Result<R, X> foldToResult(R initial, BiFunction<R, T, Result<R, X>> folder);

    <R> Option<R> foldMapping(Function<T, R> mapper, BiFunction<R, T, R> folder);

    Stream<T> filter(Predicate<T> predicate);

    Stream<T> concat(Stream<T> other);

    Option<T> next();

    boolean anyMatch(Predicate<T> predicate);
}

File: src/clang/magma/collect/stream/Stream.c
R foldWithInitial(R initial, struct BiFunction_R_T_R folder){
}
struct Stream_R map(struct Function_T_R mapper){
}
C collect(struct Collector_T_C collector){
}
struct Stream_R flatMap(struct Function_T_Stream_R mapper){
}
struct Result_R_X foldToResult(R initial, struct BiFunction_R_T_Result_R_X folder){
}
struct Option_R foldMapping(struct Function_T_R mapper, struct BiFunction_R_T_R folder){
}
struct Stream_T filter(struct Predicate_T predicate){
}
struct Stream_T concat(struct Stream_T other){
}
struct Option_T next(){
}
